src/cargo/core/source.rs

```diff
@@ -223,9 +225,39 @@ impl SourceId {
         Ok(SourceId::for_registry(&url))
     }
 
+    pub fn alt_registry(config: &Config, key: &str) -> CargoResult<SourceId> {
+        let registries = config.get_table("registries")?;
```

Comment:
Due to various bugs in the implementation today, using `get_table` will end up bypassing the environment variable fallbacks for fetching keys. Perhaps this could attempt `get_string` directly (or a similar method) to ensure that config-via-env-var works?